STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_forward_interface.c

```diff
@@ -0,0 +1,322 @@
+        remain_size_local_buffer -= size_to_copy;
+
+        // increase the passed size for the caller
+        total_size_copied += size_to_copy;
+
+        // increase offset to keep track of our position in the inner buffer
+        blob_http_cb->_internal.read_offset += size_to_copy;
+
+        // set result to AZ_OK for next loop
+        result = AZ_OK;
+      }
+
+      // All bytes in the current packet were consumed.
+      else
+      {
+        // increase current position by size of last chunk
+        ustream_forward->_internal.inner_current_position += blob_http_cb->_internal.read_offset;
+        // reset current offset
+        blob_http_cb->_internal.read_offset = 0;
+
+        // else if we have not already grabbed the last chunk, grab another
+        if (ustream_forward->_internal.inner_current_position != ustream_forward->_internal.length)
+        {
+          if ((result = _az_nx_blob_client_grab_chunk(
+                   &blob_http_cb->_internal.http_client,
+                   &blob_http_cb->_internal.packet_ptr,
+                   BLOB_CLIENT_NX_API_WAIT_TIME))
+              == AZ_ULIB_EOF)
+          {
+            result = AZ_OK;
+          }
+        }
+
+        // else this is the last buffer-full to copy, exit
+        else
+        {
+          break;
+        }
+      }
+    } while ((remain_size_local_buffer > 0) && (result == AZ_OK));
+
+    // load total size copied into returned size if successful
+    if (result == AZ_OK)
+    {
+      *size = total_size_copied;
+    }
+  }
+
+  return result;
+}
+
+static size_t concrete_get_size(az_ulib_ustream_forward* ustream_forward)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+
+  return ustream_forward->_internal.length - ustream_forward->_internal.inner_current_position;
+}
+
+static az_result concrete_dispose(az_ulib_ustream_forward* ustream_forward)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+
+  AZ_ULIB_TRY
+  {
+    // dispose of blob http control block
+    az_blob_http_cb* blob_http_cb = (az_blob_http_cb*)ustream_forward->_internal.ptr;
+    AZ_ULIB_THROW_IF_AZ_ERROR(_az_nx_blob_client_dispose(
+        &blob_http_cb->_internal.http_client, &blob_http_cb->_internal.packet_ptr));
+
+    if (ustream_forward->_internal.data_release)
+    {
+      /* If `data_release` was provided it is because `ptr` is not `const`. So, we have a
+          Warning exception here to remove the `const` qualification of the `ptr`. */
+      IGNORE_POINTER_TYPE_QUALIFICATION
+      ustream_forward->_internal.data_release(ustream_forward->_internal.ptr);
+      RESUME_WARNINGS
+    }
+    if (ustream_forward->_internal.ustream_forward_release)
+    {
+      ustream_forward->_internal.ustream_forward_release(ustream_forward);
+    }
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+AZ_NODISCARD az_result az_ulib_ustream_forward_init(
```

Comment:
"az_ulib_ustream_forward_init"
This function is not part of the public API. Please remove `az_ulib_` from the name, make it static, remove the `AZ_NODISCARD` and remove the precondition tests.

Reply:
Or just move the initialization inside of `az_blob_create_ustream_forward_from_blob` and delete this function.